### PR TITLE
User profile

### PR DESCRIPTION
**Pull request by jonaengs**
System for brukerprofil, prosjekter og ferdigheter

### CODE REVIEW

File: recommendation/forms.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+class TagForm(forms.ModelForm):
+
+    class Meta:
+        model = Tag
+        fields = ('name', )
+
+    def clean(self):
+        name = str(self.cleaned_data['name']).upper()
+        other_tags = [str(t).upper() for t in Tag.objects.all()]
+        print(name, other_tags)
```
Comment (tenstad) [inline recommendation/forms.py]: Remove print

File: recommendation/helpers.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    :param unrelated_entries: boolean -> Return entries with zero tags in common with param entry
+    :return: list of Entry objects sorted in descending order by the number of tags shared with param entry.
+    """
+    # Creates a dict with keys=Entry objects, values=no. same tags as param entry
+    tags = get_entry_tags(entry)
+    entry_tag_dict = {_entry: get_entry_tags(_entry) for _entry in Entry.objects.all()}
+    entry_tag_dict.pop(entry)  # removes param entry from list of entries to avoid matching with itself
+    for key, value in entry_tag_dict.items():  # switches the values from lists of tags to # tags in common with param
+        entry_tag_dict[key] = len(set(tags).intersection(set(value)))
+    print(entry_tag_dict)
```
Comment (tenstad) [inline recommendation/helpers.py]: Remove print

File: recommendation/helpers.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    # creates the list mentioned as :return
```
Comment (tenstad) [inline recommendation/helpers.py]: What is the difference between if and else branch?
Better to use sorted()?

File: recommendation/urls.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    EditEntryView, view_entry, CreateTagView, ListEntriesView
+
+urlpatterns = [
+    url(r'^$', ListEntriesView.as_view(), name="entry_list"),
+    url(r'^create/$', CreateEntryView.as_view(), name="entry_form"),
+    url(r'^edit/(?P<pk>\d+)/$', EditEntryView.as_view(), name="edit_entry"),
+    url(r'^delete/(?P<pk>\d+)/$', DeleteEntryView.as_view(), name="delete_entry"),
+    url(r'^detail/(?P<pk>\d+)/$', view_entry, name="entry_detail"),
+    url(r'^tag/add/$', CreateTagView.as_view(), name="tag_form"),
+]
```
Comment (tenstad) [inline recommendation/urls.py]: Easier to use path

File: recommendation/views.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+from .helpers import get_related_entries, get_entry_tags
+from .forms import EntryForm, TagForm
+from .models import Entry, Tag
+
+
+class ListEntriesView(ListView):
+    model = Entry
+
+    def get_queryset(self):
+        return Entry.objects.all().order_by('-creation_date')  # minus-sign => descending order
```
Comment (tenstad) [inline recommendation/views.py]: Possible to define order in model and avoid whole view by passing model as argument to ListView in urls?

File: recommendation/views.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+class EditEntryView(PermissionRequiredMixin, UpdateView):
+    redirect_field_name = 'recommend/entry_detail.html'
+    permission_required = 'recommendation.change_entry'
+    success_url = reverse_lazy('entry_list')
+    model = Entry
+    form_class = EntryForm
+
+
+class DetailEntryView(DetailView):
```
Comment (tenstad) [inline recommendation/views.py]: Remove view and pass model as argument to DetailView in urls

File: user_profile/urls.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    url(r'project/(?P<pk>\d+)/delete/$', DeleteProjectView.as_view(), name='delete_project'),
+    url(r'project/(?P<pk>\d+)/edit/$', EditProjectView.as_view(), name='edit_project'),
+    url(r'project/(?P<pk>\d+)/apply/$', apply_to_project, name='apply_to_project'),
+    url(r'project/(?P<pk>\d+)$', DetailProjectView.as_view(), name='project'),
+    url(r'project/new/$', CreateProjectView.as_view(), name='project_form'),
+    url(r'skill/(?P<pk>\d+)$', DetailSkillView.as_view(), name='skill'),
+
+    # must always be at the end, else it will interpret skill, project etc. as usernames
+    url(r'(?P<username>[-\w.]+)/$', profile, name='profile'),
+    ]
```
Comment (tenstad) [inline user_profile/urls.py]: Use path

File: user_profile/views.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    model = Skill
+
+
+class DetailProjectView(DetailView):
+    model = Project
+
+
+class ListProjectView(ListView):
+    model = Project
+
```
Comment (tenstad) [inline user_profile/views.py]: All above: remove view and pass model as argument in urls

File: user_profile/views.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+def reject_applicant(request, pk, username):
+    return manage_applicant(request, pk, username, False)
+
+
+def apply_to_project(request, pk):
+    if request.method == 'POST':
+        user = get_object_or_404(User, username=request.user.username)
+        project = get_object_or_404(Project, pk=pk)
+
+        # if not rejected, already an applicant or already a member
```
Comment (tenstad) [inline user_profile/views.py]: filter.exists might be faster
Not sure if Django queries all from database then checks for membership or actually does the membership check as a database query.

File: user_profile/views.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    user = get_object_or_404(User, username=username)
+    try:
+        user.profile  # Accessing a non-existent profile (they do no exist by default) triggers an error
+    except Profile.DoesNotExist:
+        new_profile = Profile(user=user)
+        new_profile.save()
+
+    # populate()
+    form = ProfileForm()
+    print(user.profile.picture.name)
```
Comment (tenstad) [inline user_profile/views.py]: Remove print

File: user_profile/views.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        return HttpResponseRedirect(reverse('project', kwargs={'pk': pk}))
+
+    return HttpResponseRedirect('/')
+
+
+def profile(request, username):
+    user = get_object_or_404(User, username=username)
+    try:
+        user.profile  # Accessing a non-existent profile (they do no exist by default) triggers an error
+    except Profile.DoesNotExist:
```
Comment (tenstad) [inline user_profile/views.py]: Profile.objects.create(user=user)

File: user_profile/views.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        new_profile.save()
+
+    # populate()
+    form = ProfileForm()
+    print(user.profile.picture.name)
+    if request.method == 'POST':
+        form = ProfileForm(request.POST, request.FILES)
+        if form.is_valid():
+            # user.profile.skills.remove(user.profile.skills.all()) seems to delete all skills completely from the DB
+            # removing the connection one way does not remove it the other way...?
```
Comment (tenstad) [inline user_profile/views.py]: Does clear work?
https://docs.djangoproject.com/en/2.0/ref/models/relations/#django.db.models.fields.related.RelatedManager.clear

File: vim.exe.stackdump [hunk 13]
```diff
@@ -0,0 +1,19 @@
+Stack trace:
```
Comment (tenstad) [inline vim.exe.stackdump]: Delete file
Maybe add to .gitignore

File: recommendation/views.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
+from pprint import pprint
```
Comment (tenstad) [inline recommendation/views.py]: Unused?

File: user_profile/views.py (hunk 9 above)
Comment (jonaengs) [reply user_profile/views.py]: According to what I've read on SO, get_or_404(<Object>, <key>) works like a <Object>.get(<key>) with a raise 404Error in-built if nothing is found.

File: user_profile/views.py (hunk 11 above)
Comment (jonaengs) [reply user_profile/views.py]: Done. Doesn't show here for some reason.

**Pull request merged**